### PR TITLE
Add simulator

### PR DESCRIPTION
**Pull request by wbernoudy**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -48,10 +48,16 @@ def build_extensions(self):
...
+        sources=["dwave/gate/simulator/simulator.pyx"],
+        include_dirs=["dwave/gate/simulator/"],
+        language="c++",
+    ),
+    Extension(
+        name="dwave.gate.registers.cyregister",
+        sources=["dwave/gate/registers/cyregister.pyx"],
         include_dirs=["dwave/gate/registers/"],
-        language='c++',
+        language="c++",
```
Comment (arcondello) [inline setup.py:60]: it's better to just put ```
# distutils: language = c++
# cython: language_level=3
```
at the top of your pxd files

**Review by arcondello — APPROVED** [review]
LGTM

File: dwave/gate/simulator/operation_generation.py [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+
+
+def compile_gate(
+    gate_matrix: Union[np.ndarray, List[List[EntryType]]]
+) -> Tuple[Set[int], Dict[int, List[Tuple[int, EntryType]]]]:
+    """Compile a given gate into a minimal set of instructions.
+
+    Args:
+        gate_matrix: Square matrix representing the gate. Values may be scalars or C variables that
+        refer to floats.
```
Comment (thisac) [inline dwave/gate/simulator/operation_generation.py:37]:
```suggestion
        gate_matrix: Square matrix representing the gate. Values may be scalars or C variables that
            refer to floats.
```

**Review by thisac — APPROVED** [review]
Thanks @wbernoudy. Looks great!

**Comment by randomir** [comment]
Nice work. Cool use of `cgen`!

**Pull request merged**